src/patchkernel/volume_mapper.cpp

```diff
@@ -0,0 +1,207 @@
+/*---------------------------------------------------------------------------*\
+ *
+ *  bitpit
+ *
+ *  Copyright (C) 2015-2019 OPTIMAD engineering Srl
+ *
+ *  -------------------------------------------------------------------------
+ *  License
+ *  This file is part of bitpit.
+ *
+ *  bitpit is free software: you can redistribute it and/or modify it
+ *  under the terms of the GNU Lesser General Public License v3 (LGPL)
+ *  as published by the Free Software Foundation.
+ *
+ *  bitpit is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+ *  License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with bitpit. If not, see <http://www.gnu.org/licenses/>.
+ *
+\*---------------------------------------------------------------------------*/
+
+#include <cassert>
+
+#include "volume_mapper.hpp"
+
+namespace bitpit {
+
+/**
+ * \class VolumeMapper
+ * \ingroup Ref
+ *
+ * \brief The VolumeMapper is the class to map two meshes.
+ *
+ * The VolumeMapper allows to map meshes of class VolumeKernel. The meshes are defined as
+ * a reference mesh and a mapped mesh.
+ * The object can provide a direct mapper and a inverse mapper between only the Cells of the meshes.
+ * The information given by a mapper object is analogous to an adaptation info to adapt the mapped mesh
+ * to the reference one (or vice versa in case of inverse mapper).
+ * The two meshes have to be imperatively linked at declaration of the mapper object.
+ * To compute the mapper the first time call initialize method.
+ * Then, if the reference mesh OR the mapped mesh (one at a time) is adapted,
+ * the mapper can be adapted together by passing the adaptation information to the
+ * prepare and alter methods. A load-balancing procedure is not allowed, i.e. the mapper
+ * has to be entirely recomputed after a load-balance of a mesh.
+ *
+ * Note that the VolumeMapper class is a pure virtual class, only objects of its complete derived classes can
+ * be instantiated.
+ * The core method of the mapping procedure and the adapting mesh functions have to be implemented in a derived class.
+ *
+ * The VolumeMapper allows to map meshes of class: VolOctree (see derived MapperVolOctree class).
+ *
+ */
+
+/**
+ * Creates a new VolumeMapper object.
+ */
+# if BITPIT_ENABLE_MPI
+/**
+ * \param[in] comm The MPI communicator used by the Ref object. MPI_COMM_WORLD is the default value.
+ */
+VolumeMapper::VolumeMapper(MPI_Comm comm)
+# else
+VolumeMapper::VolumeMapper()
+# endif
+: m_referenceMesh(nullptr), m_mappedMesh(nullptr), m_mapper(1)
+{
+
+#if BITPIT_ENABLE_MPI
+    m_communicator = MPI_COMM_NULL;
+    initializeCommunicator(comm);
+    MPI_Comm_size(m_communicator, &m_nProcs);
+    MPI_Comm_rank(m_communicator, &m_rank);
+    m_rank = 0;
+    m_nProcs = 1;
```

Comment:
@edoardolombardi It seems strange that the values of m_rank and m_nProcs are overwritten just after being set.